hydra/hydra.go

```diff
@@ -192,7 +205,73 @@ func NewHydra(ctx context.Context, options Options) (*Hydra, error) {
 		go handleBootstrapStatus(hdCtx, bsCh)
 
 		hds = append(hds, hd)
+
+		// dial back hooks
+		evts := []interface{}{
+			new(event.EvtPeerIdentificationCompleted),
+		}
+
+		subs, err := hd.Host.EventBus().Subscribe(evts, eventbus.BufSize(256))
+		if err != nil {
+			return nil, fmt.Errorf("head could not subscribe to eventbus events; err: %w", err)
+		}
+		go func(hd *head.Head, subs event.Subscription) {
+			defer subs.Close()
+			for {
+				select {
+				case evt := <-subs.Out():
+					ev := evt.(event.EvtPeerIdentificationCompleted)
+					p := ev.Peer
+					addrs := hd.Host.Peerstore().Addrs(p)
+
+					// dial back on quic if peer advertises a quic address
+					for _, a := range addrs {
+						if mafmt.QUIC.Matches(a) {
+							nQUICConns.Add(1)
+
+							if err := hd.QuicDialBackHost.Connect(hdCtx, peer.AddrInfo{ID: p}); err == nil {
+								nQUICDialbacks.Add(1)
+
+								// close the connection as we don't need it anymore
+								if err := hd.QuicDialBackHost.Network().ClosePeer(p); err != nil {
+									fmt.Fprintf(os.Stderr, "\n quic dial back: failed to close connection to peer %s, err: %s", p.Pretty(),
+										err)
+								}
+							}
+
+							pct := float64(100) * float64(nQUICDialbacks.Load()) / float64(nQUICConns.Load())
+							stats.Record(hdCtx, metrics.QUICDialBackPct.M(pct))
+							break
+						}
+					}
+
+					// dial back on tcp if peer advertises a tcp address
+					for _, a := range addrs {
+						if mafmt.TCP.Matches(a) {
+							nTCPConns.Add(1)
+
+							if err := hd.TcpDialBackHost.Connect(hdCtx, peer.AddrInfo{ID: p}); err == nil {
```

Comment:
Should we log the error if one occurs? Or have a sanity check that this is a timeout error (`net.Error` with `Timeout() == true`?

Reply:
The only other error I could imagine is a QUIC version negotiation error, which should be rare, as we've deployed draft-29 for quite a while now. If there are other errors, we should probably look into that.

Reply:
Hmmm... are you sure that dial errors because of NATS would all be timeout errors ? Aren't there NATs out there that send ICMP messages for unsolicited requests ?

Reply:
Hadn't thought of ICMP. In UDP, you only get ICMP messages for connected sockets (which we don't use), not sure how this works in TCP. Actually, it would actually be an interesting thing to know how those dials fail.

Another reason I'm asking for error logging is that I recently went through the QUIC error logs and discovered that a lot of dials were failing when connection to our old Bootstrap nodes. We managed to make sense of this, but it was unexpected at first, so I suggest that there's value in logging these errors (or maybe exporting the error string to Grafana if we want a single source of truth? Not sure how to best do this).

Reply:
@Stebalien Would you know what we'd see in the logs for these dial failures if the peers were NATT'd ? My only concerns is bloating the logs because large parts of the network are still undialable.

Reply:
Really, you can get any error. You might get a TCP reset from the NAT, you might get an ICMP rejection, you might get a timeout.